### DISCUSSION TITLE
Side project - todayspokemon.com

### DISCUSSION THREAD

**Discussion by joshchagani** [statelyai/xstate#2564] [Show and tell]
Howdy Friends,

I did a little project that consists of two Lambdas written in Node and the FE written in React. Here's the project in the wild [todayspokemon.com](https://todayspokemon.com) and here's the [Github repo](https://github.com/joshchagani/pokemon-a-day). 

I have a couple questions concerning passing full context and best other best practices:

- I have an AWS Lambda as my REST API. It connects to my DB and then sends the goods back to the client. I need the entire context within XState to be sent but the only way that I could access the values within the context was to destructure every single property. Lines 54 - 66 [on my Github project](https://github.com/joshchagani/pokemon-a-day/blob/main/server/rest-pokemon-api/app/handler.ts) show what I mean. Is there a way to get the entire context without pulling out the values individually? Would that be an anti-pattern with XState since being explicit seems to be a foundation? 

- Second question deals with the FE. I don't want to ask something incredibly vague like, "Am I doing this right?" but it's how I'm feeling. Here's the [bulk of the FE machine](https://github.com/joshchagani/pokemon-a-day/blob/main/client/src/machine/dataMachine.tsx). I feel like I rely on "invoke" to progress the machine where at times maybe I should be spawning machines? Would "entry" be a better use? Should all functions be written within the "actions" of "createMachine"? 

Thanks for taking the time to checkout my project!

**Comment by Andarist** [comment]
> I need the entire context within XState to be sent but the only way that I could access the values within the context was to destructure every single property. Lines 54 - 66 on my Github project show what I mean. Is there a way to get the entire context without pulling out the values individually?

This shouldn't be a problem. A simple line like this should just work:
```js
const { pokemonInfo } = state.context
```

If you experience any problem with that - please create a runnable repro case that we could easily inspect to see what happens.

> Second question deals with the FE. I don't want to ask something incredibly vague like, "Am I doing this right?" but it's how I'm feeling. Here's the bulk of the FE machine.

In cases like this, it would be nice if you could share this as a machine saved in our visualizer: https://stately.ai/viz . It would make taking a quick look at it much easier for us.

> I feel like I rely on "invoke" to progress the machine where at times maybe I should be spawning machines?

If your flow is sequential then invoke it's the most straightforward thing to do - because it allows you to easily listen for the "subtask"'s completion using `onDone`.

> Would "entry" be a better use?

Actions (entry included) are mostly treated as fire & forget. If you need to communicate with the parent (the caller) then it's easier to achieve this using invoke (or spawn)

> Should all functions be written within the "actions" of "createMachine"?

It depends on your preference but using `actions` in the `createMachine`'s second argument makes it easier to serialize the machine, visualize it etc. It sometimes might feel like a chore though if you don't rely heavily on those things so choose whatever works for you. It's always easy to migrate to the other "style".
